cachecontrol/heuristics.py

```diff
@@ -77,3 +78,36 @@ def update_headers(self, response):
     def warning(self, response):
         tmpl = '110 - Automatically cached for %s. Response might be stale'
         return tmpl % self.delta
+
+_ONE_DAY_IN_SECONDS = timedelta(days=1)
+
+class LastModifiedHeuristic(BaseHeuristic):
+    """
+    Apply the heuristic suggested by RFC 7234, 4.2.2,
+    using the Last-Modified date, when no explicit
+    freshness is specified.
+    """
+
+    def update_headers(self, response):
+        if 'expires' not in response.headers:
+            if 'cache-control' not in response.headers or response.headers['cache-control'] == 'public':
+                # RFC 7231, 6.1
+                if response.status in [200, 203, 204, 206, 300, 301, 404, 405, 410, 414, 501]:
+                    if 'last-modified' in response.headers:
+                        last_modified = parsedate(response.headers['last-modified'])
+                        now = datetime.now()
+                        age = now - datetime(*last_modified[:6])
+                        expires = now + (age / 10)
+
+                        return {'expires': datetime_to_header(expires)}
+        return {}
+
+    def warning(self, response):
+        now = datetime.now()
+        date = parsedate(response.headers['date'])
+        current_age = max(timedelta(), now - datetime(*date[:6]))
+
+        if current_age > _ONE_DAY_IN_SECONDS:
+            return '113 - Heuristic Expiration'
```

Comment:
There's a bug here. This logic only runs when the resource is fetched, at which point it'll still be heuristically fresh. It needs to run when the resource is fetched from the cache, but the cached headers at that point include the synthetic `Expires`, indistinguishable from a real one.

I can see a few options;

* Unconditionally set the warning.
* Always store the original headers, and re-run `update_headers` when determining freshness
* Store an extra flag with the cached response indicating that a warning should be calculated if necessary.